dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,634 @@
+    return boundary::idx_clamp(idx, 0, len);
+  }
+};
+
+struct Wrap {
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len) const {
+    assert(len > 0);
+    return boundary::idx_wrap(idx, len);
+  }
+};
+
+template <typename Remap, typename In>
+struct InLoaderBorderRemap : protected Remap {
+  DALI_HOST_DEV DALI_FORCEINLINE int remap_height(int idx, const ShapeDesc& sample_shape) const {
+    return this->border_remap(idx, sample_shape.h);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int remap_width(int idx, const ShapeDesc& sample_shape) const {
+    return border_remap_strided(idx, sample_shape.w, sample_shape.c, sample_shape.wc);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE In load(const In* __restrict__ in, int y, int x,
+                                         const ShapeDesc& sample_shape) const {
+    return in[y * static_cast<int64_t>(sample_shape.wc) + x];
+  }
+
+ protected:
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap_strided(int idx, int reflect_dim_size,
+                                                          int inner_stride,
+                                                          int total_stride) const {
+    if (idx < 0) {
+      int reflect_dim_idx = (idx + 1) / inner_stride - 1;
+      int inner_dim_idx = (idx + 1) % inner_stride + inner_stride - 1;
+      return this->border_remap(reflect_dim_idx, reflect_dim_size) * inner_stride + inner_dim_idx;
+    }
+    if (idx >= total_stride) {
+      return this->border_remap(idx / inner_stride, reflect_dim_size) * inner_stride +
+             idx % inner_stride;
+    }
+    return idx;
+  }
+};
+
+template <typename In>
+struct InLoaderPad {
+  DALI_HOST_DEV DALI_FORCEINLINE int remap_height(int idx, const ShapeDesc& sample_shape) const {
+    return idx;
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int remap_width(int idx, const ShapeDesc& sample_shape) const {
+    return idx;
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE In load(const In* __restrict__ in, int y, int x,
+                                         const ShapeDesc& sample_shape) const {
+    if (y < 0 || x < 0 || x >= sample_shape.wc || y >= sample_shape.h) {
+      return fill_value;
+    }
+    return in[y * static_cast<int64_t>(sample_shape.wc) + x];
+  }
+
+  In fill_value;
+};
+
+template <typename InLoader>
+struct InLoaderFactory {
+  using T = InLoader;
+  DALI_DEVICE DALI_FORCEINLINE T& operator()(int sample_idx) {
+    return in_loader;
+  }
+  T in_loader;
+};
+
+template <typename In>
+struct InLoaderFactory<InLoaderPad<In>> {
+  using T = InLoaderPad<In>;
+  DALI_DEVICE DALI_FORCEINLINE T operator()(int sample_idx) {
+    if (fill_values == nullptr) {
+      return {0};
+    }
+    return {fill_values[sample_idx][0]};
+  }
+  const In** fill_values;
+};
+
+/** @} */  // end of InputLoader
+
+
+/** @defgroup InputConv The specializations provide ``compute`` method that computes convolution
+ * for the ``block_width x lanes`` patch and stores it in the provided ``acc``.
+ * @{
+ */
```

Comment:
Can you add the docstrings for the `stride_grid` and other implementation parts? I wouldn't oppose putting a sentence about the traversal order that you shared today (warp with threadIdx.x mapping to the width of the kernel, and lanes to the height of the kernel while being unrolled), but maybe I'm a bit overly eager for commenting stuff recently

Reply:
I've added comment about the general idea from the slide at the top-level `__global__ filter`